### DISCUSSION TITLE
Compilação, IL e CLR

### DISCUSSION THREAD

**Discussion by ghost** [balta-io/2801#67] [Perguntas e Respostas] [answered]
No Módulo C# - Aula IL, o slide 1 afirma que o CLR compila qualquer linguagem da Microsoft (C#, VB.NET, etc.) para IL. 

Entretanto, as aulas anteriores me fizeram entender que cada linguagem tinha um compilador para IL independente do CLR. 

Isto significa que o Gerenciador de Linguagem Comum carrega consigo, além do compilador JIT de IL para binário, todos os compiladores das linguagens .NET?

**Comment by andrebaltieri** [comment, marked as answer]
Boa noite, @PolarisTDM, como vai?

Muito obrigado pelo feedback, se puder me dizer mais detalhes sobre quais aulas ou quais trechos que te confundiram, vou tentar melhorar a didática nelas...

Sobre sua pergunta, eles carregam uma espécie de tradutor da linguagem para o IL, e o CLR sim tem o compilador do IL para binário.

**Comment by ghost** [comment]
Tudo ótimo, Balta. Tenho gostado bastante da tua plataforma. Cada aula é geralmente curta, mas passo uns bons minutos anotando e absorvendo os conceitos de cada uma. Isto deve significar que você é cirúrgico na escolha dos temas de cada uma.

### Sobre os detalhes

**Módulo C# - Aula "Compilação e Gerenciamento" - Minuto 2:19** 
- Você fala basicamente que cada linguagem da Microsoft tem seu tradutor, de código-fonte para IL, separado do CLR.

**Módulo C# - Aula "IL" - Minuto 00:46**
- Já na aula posterior, você fala que o CLR tem as duas responsabilidades: a de traduzir o código-fonte (de qualquer linguagem suportada pelo .NET) e a de gerenciar execução de IL.

_A primeira aula me fez entender o seguinte: Há compiladores diferentes para cada linguagem da Microsoft & estes compiladores são isolados/independentes da CLR, a qual fica somente responsável por compilar, gerenciar/executar IL._

_Já a segunda aula me fez entender que: Há compiladores diferentes para cada linguagem da Microsoft & eles já estão embutidos no CLR._

Os dois entendimentos estão corretos?

Abraço.

**Reply by andrebaltieri** [reply]
Perfeito, muito obrigado pelo auxílio, vou elaborar melhor estas aulas... só fico com um pouco de receio de assustar quem está começando... por isto não detalho tanto... mas vamos lá...

Esta imagem ilustra bem como as coisas acontecem:
![Compilação do C#](https://freecontent.manning.com/wp-content/uploads/how-is-c-compiled_01.png)

O JIT age entre ambas partes, ele faz toda esta interação, mas como a primeira parte é mais uma "tradução" normalmente não chamamos de "compilar".

Aqui tem mais detalhes técnicos sobre isto:
https://docs.microsoft.com/pt-br/dotnet/standard/clr

Se tiver mais dúvidas, manda aqui, gosto bastante do assunto :)
